Dynamic module resolution in socket.io < 3 prevents use with Webpack

I am indirectly using koa-socket-2 via https://github.com/boardgameio/boardgame.io. 
Unfortunately, socket.io before version 3.0.0 (koa-socket-2 requires 2.3.0) has a bug which prevents it from being used with TypeScript and Webpack (not sure which one causes the error). Compilation already generates a warning:

```
WARNING in ./node_modules/koa-socket-2/node_modules/socket.io/lib/index.js 113:11-32
Critical dependency: the request of a dependency is an expression
 @ ./node_modules/koa-socket-2/index.js 4:17-39
 @ ./node_modules/boardgame.io/dist/cjs/server.js 13:25-48
 @ ./src/server.ts 1:0-45 3:15-21
```

Execution subsequently results in the following message:

```
webpack://[redacted]/./node_modules/koa-socket-2/node_modules/socket.io/lib/_sync?:4
        throw e;
        ^
Error: Cannot find module 'socket.io-client/dist/socket.io.js'
    at Function.webpackEmptyContext [as resolve] (webpack://[redacted]/./node_modules/koa-socket-2/node_modules/socket.io/lib/_sync?:2:10)
    at resolvePath (webpack://[redacted]/./node_modules/koa-socket-2/node_modules/socket.io/lib/index.js?:113:125)
    at Server.serveClient (webpack://[redacted]/./node_modules/koa-socket-2/node_modules/socket.io/lib/index.js?:116:25)
    at new Server (webpack://[redacted]/./node_modules/koa-socket-2/node_modules/socket.io/lib/index.js?:53:8)
    at IO.attach (webpack://[redacted]/./node_modules/koa-socket-2/index.js?:132:15)
    at SocketIO.init (webpack://[redacted]/./node_modules/boardgame.io/dist/cjs/server.js?:3163:12)
    at Server (webpack://[redacted]/./node_modules/boardgame.io/dist/cjs/server.js?:3264:15)
    at eval (webpack://[redacted]/./src/server.ts?:6:75)
    at Module../src/server.ts ([redacted]/dist/server.js:31834:1)
    at __webpack_require__ ([redacted]/dist/server.js:32492:42) {
  code: 'MODULE_NOT_FOUND'
```

The cause is a dynamic import which is used by the socket.io server to serve the compiled client code too (see https://github.com/socketio/socket.io/blob/47161a65d40c2587535de750ac4c7d448e5842ba/lib/index.js#L113). The dynamic import cannot be correctly handled by Webpack which is why the imported file is not included in the bundle.

I haven't tested it yet but it looks like the issue has been accidentally resolved through this commit:

https://github.com/socketio/socket.io/commit/7603da71a535481e3fc60e38b013abf78516d322#diff-d5a4b3a0309f2337144861084c1014523cced561eb722cb9684c63f5e41e0567L228

For koa-socket-2 I see two possible solutions to this issue: Either to upgrade socket.io to version 3.0 or to disable to serve the client code, which can be achieved by importing socket.io as follows:

`const io = require('socket.io')(server, {
  serveClient: false,
});`

Would the second solution be feasible for koa-socket-2 or do users depend on the client code being served through socket.io?

Sounds good.  Let’s get it upgraded!

> On Nov 15, 2020, at 3:30 PM, Jonas Lieb <notifications@github.com> wrote:
> 
> 
> I am indirectly using koa-socket-2 via https://github.com/boardgameio/boardgame.io <https://github.com/boardgameio/boardgame.io>.
> Unfortunately, socket.io before version 3.0.0 (koa-socket-2 requires 2.3.0) has a bug which prevents it from being used with TypeScript and Webpack (not sure which one causes the error). Compilation already generates a warning:
> 
> WARNING in ./node_modules/koa-socket-2/node_modules/socket.io/lib/index.js 113:11-32
> Critical dependency: the request of a dependency is an expression
>  @ ./node_modules/koa-socket-2/index.js 4:17-39
>  @ ./node_modules/boardgame.io/dist/cjs/server.js 13:25-48
>  @ ./src/server.ts 1:0-45 3:15-21
> Execution subsequently results in the following message:
> 
> webpack://[redacted]/./node_modules/koa-socket-2/node_modules/socket.io/lib/_sync?:4
>         throw e;
>         ^
> Error: Cannot find module 'socket.io-client/dist/socket.io.js'
>     at Function.webpackEmptyContext [as resolve] (webpack://[redacted]/./node_modules/koa-socket-2/node_modules/socket.io/lib/_sync?:2:10)
>     at resolvePath (webpack://[redacted]/./node_modules/koa-socket-2/node_modules/socket.io/lib/index.js?:113:125)
>     at Server.serveClient (webpack://[redacted]/./node_modules/koa-socket-2/node_modules/socket.io/lib/index.js?:116:25)
>     at new Server (webpack://[redacted]/./node_modules/koa-socket-2/node_modules/socket.io/lib/index.js?:53:8)
>     at IO.attach (webpack://[redacted]/./node_modules/koa-socket-2/index.js?:132:15)
>     at SocketIO.init (webpack://[redacted]/./node_modules/boardgame.io/dist/cjs/server.js?:3163:12)
>     at Server (webpack://[redacted]/./node_modules/boardgame.io/dist/cjs/server.js?:3264:15)
>     at eval (webpack://[redacted]/./src/server.ts?:6:75)
>     at Module../src/server.ts ([redacted]/dist/server.js:31834:1)
>     at __webpack_require__ ([redacted]/dist/server.js:32492:42) {
>   code: 'MODULE_NOT_FOUND'
> The cause is a dynamic import which is used by the socket.io server to serve the compiled client code too (see https://github.com/socketio/socket.io/blob/47161a65d40c2587535de750ac4c7d448e5842ba/lib/index.js#L113 <https://github.com/socketio/socket.io/blob/47161a65d40c2587535de750ac4c7d448e5842ba/lib/index.js#L113>). The dynamic import cannot be correctly handled by Webpack which is why the imported file is not included in the bundle.
> 
> I haven't tested it yet but it looks like the issue has been accidentally resolved through this commit:
> 
> socketio/socket.io@7603da7#diff-d5a4b3a0309f2337144861084c1014523cced561eb722cb9684c63f5e41e0567L228 <https://github.com/socketio/socket.io/commit/7603da71a535481e3fc60e38b013abf78516d322#diff-d5a4b3a0309f2337144861084c1014523cced561eb722cb9684c63f5e41e0567L228>
> For koa-socket-2 I see two possible solutions to this issue: Either to upgrade socket.io to version 3.0 or to disable to serve the client code, which can be achieved by importing socket.io as follows:
> 
> const io = require('socket.io')(server, { serveClient: false, });
> 
> Would the second solution be feasible for koa-socket-2 or do users depend on the client code being served through socket.io?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub <https://github.com/ambelovsky/koa-socket-2/issues/21>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAMURJTG5ISORNQAWMK5KKLSQAM3TANCNFSM4TWLFJUQ>.
>

--- a/README.md
+++ b/README.md
@@ -5,7 +5,7 @@
 
 > Sugar for connecting socket.io to a Koa instance
 
-**Koa-socket-2 uses socket.io v2.  It is recommended that you connect to a koa-socket-2 server with a socket.io v2 client.**
+**Koa-socket-2 uses socket.io v3.  It is recommended that you connect to a koa-socket-2 server with a socket.io v3 client.**
 
 Koa-socket-2 is only compatible with Koa v2 style of middleware (where context is passed as a parameter).
 

--- a/index.js
+++ b/index.js
@@ -129,7 +129,7 @@ module.exports = class IO {
       throw new Error( 'Default namespace can not be hidden' );
     }
 
-    app._io = new socketIO( app.server, this.opts.ioOptions );
+    app._io = socketIO( app.server, this.opts.ioOptions );
 
     if ( this.opts.namespace ) {
       this.attachNamespace( app, this.opts.namespace );

--- a/package.json
+++ b/package.json
@@ -37,6 +37,7 @@
     "coveralls": "^3.0.0",
     "istanbul": "^0.4.5",
     "istanbul-harmony": "^0.3.16",
+    "glob": "^7.1.6",
     "koa": "^2.7.0",
     "minimist": "^1.2.0",
     "socket.io-client": "^2.0.1",

--- a/package.json
+++ b/package.json
@@ -5,8 +5,8 @@
   "main": "index.js",
   "scripts": {
     "test": "node spec/index.js spec/*.test.js",
-    "cover": "istanbul cover --report lcovonly --print detail spec/index.js spec/*.test.js",
-    "travis": "npm run cover && cat coverage/lcov.info | coveralls",
+    "cover": "nyc npm run test",
+    "travis": "npm run cover && nyc report --reporter=text-lcov | coveralls",
     "example": "node example/server"
   },
   "keywords": [
@@ -35,11 +35,10 @@
   },
   "devDependencies": {
     "coveralls": "^3.0.0",
-    "istanbul": "^0.4.5",
-    "istanbul-harmony": "^0.3.16",
     "glob": "^7.1.6",
     "koa": "^2.7.0",
     "minimist": "^1.2.0",
+    "nyc": "^15.1.0",
     "socket.io-client": "^2.0.1",
     "tape": "^4.6.3"
   }

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "koa-socket-2",
-  "version": "1.2.0",
+  "version": "2.0.0",
   "description": "Koa meets socket.io connected socket",
   "main": "index.js",
   "scripts": {

--- a/package.json
+++ b/package.json
@@ -31,7 +31,7 @@
   "license": "MIT",
   "dependencies": {
     "koa-compose": "^4.1.0",
-    "socket.io": "^2.2.0"
+    "socket.io": "^3.0.2"
   },
   "devDependencies": {
     "coveralls": "^3.0.0",
@@ -39,7 +39,7 @@
     "koa": "^2.7.0",
     "minimist": "^1.2.0",
     "nyc": "^15.1.0",
-    "socket.io-client": "^2.0.1",
+    "socket.io-client": "^3.0.2",
     "tape": "^4.6.3"
   }
 }

